packages/playground/sync/src/index.ts

```diff
@@ -1,8 +1,59 @@
-export { recordFSOperations, replayFSOperations } from './fs';
-export {
-	idOffset,
-	SQLQueryMetadata,
-	recordSQLQueries,
-	replaySQLQueries,
-} from './sql';
-export { ParentWindowTransport } from './transports';
+import { PlaygroundClient } from '@wp-playground/remote';
+import { installSqlSyncMuPlugin, overrideAutoincrementSequences } from './sql';
+import { recordFSOperations, replayFSOperations } from './fs';
+import { SQLQueryMetadata, recordSQLQueries, replaySQLQueries } from './sql';
+import { PlaygroundSyncTransport, TransportMessage } from './transports';
+import { debounce } from './utils';
+import { FilesystemOperation } from '@php-wasm/universal';
+
+export interface Logger {
+	log(...args: any[]): void;
+}
+
+export interface SyncOptions {
+	autoincrementOffset: number;
+	transport: PlaygroundSyncTransport;
+	logger?: Logger;
+}
+
+export async function setupPlaygroundSync(
+	playground: PlaygroundClient,
+	{ autoincrementOffset, transport, logger }: SyncOptions
+) {
+	await installSqlSyncMuPlugin(playground);
+	await overrideAutoincrementSequences(playground, autoincrementOffset);
+
+	transport.onChangesReceived(async (changes) => {
+		logger?.log(`Received changes`, changes);
+		for (const { scope, details } of changes) {
+			if (scope === 'fs') {
+				await replayFSOperations(playground, details);
+			} else if (scope === 'sql') {
+				await replaySQLQueries(playground, details);
+			}
+		}
+	});
+
+	let localChanges: TransportMessage[] = [];
+	const debouncedFlush = debounce(() => {
+		logger?.log(`Sending changes`, localChanges);
+		transport.sendChanges(localChanges);
+		localChanges = [];
+	}, 3000);
+
+	recordSQLQueries(playground, (queries: SQLQueryMetadata[]) => {
+		localChanges.push({ scope: 'sql', details: queries });
+		debouncedFlush();
+
+		// Track autoincrement values like this:
+		// for (const query of queries) {
+		// 	if (query.subtype === 'reconstruct-insert') {
+		// 		storedIds[query.table_name] = query.last_insert_id;
+		// 	}
+		// }
+	});
+	recordFSOperations(playground, (ops: FilesystemOperation[]) => {
+		localChanges.push({ scope: 'fs', details: ops });
+		debouncedFlush();
+	});
```

Comment:
we shouldn't need to call `debouncedFlush()` manually, and that may end up adding needless overhead on actions that happen all the time, even though it should be cheap.

I think we can also have the flush perform its own timing, as an independent actor.

```js
const flushJournal = () => {
    if (!localChanges.length) {
		return;
	}
	logger?.log(`Sending changes ${localChanges}`);
	transport.sendChanges(localChanges);
	localChanges = [];
}

const loopAfterInterval = ( f, ms ) => {
    f();
    setTimeout( loopAfterInterval, ms, f, ms );
}

loopAfterInterval( flushJournal, 3000 );
```

now this only attempts to send changes every 3000ms since the last flush and when the event loop opens up to allow it, or if we want to call it at a specific time to "force sync." not a meaningful performance change, but maybe something to consider for separation of duties.

Reply:
I like it, thanks! I just rewrote it like this